lib/controller-inputs/DigitalInputGroup.cpp

```diff
@@ -0,0 +1,148 @@
+#include "DigitalInputGroup.h"
+
+DigitalInputGroup::DigitalInputGroup(BusIn& inputs, uint8_t numPins, InputDebounceType debounceType,
+                                     uint16_t integratingDebounceThres, uint16_t delayDebouceTimeMS)
+    : m_inputs(inputs),
+      m_numPins(numPins),
+      m_integratingDebounceThres(integratingDebounceThres),
+      m_delayDebounceTimeMS(delayDebouceTimeMS),
+      m_debounceType(debounceType) {
+  // To check constructor parameter at compile time involves using template, don't think that is appropriate here ...
+  // So we kind of have to fail this at runtime...it would fail anyways later on if numPins is too large
+  MBED_ASSERT(numPins <= MAX_INPUTS_NUM && numPins != 0);
+
+  m_timer.start();
+}
+
+uint8_t DigitalInputGroup::getNumInputs() const {
+  return m_numPins;
+}
+
+/* It might be a better architecture to separate out each debounce type into separate classes,
+   but I think the complexity in this class right now is still manageable and it is unlikedly
+   that we add new types of debouncing.
+*/
+void DigitalInputGroup::read() {
+  switch (m_debounceType) {
+    case InputDebounceType::NONE:
+      readNoDebounce();
+      break;
+
+    case InputDebounceType::DELAY:
+      readDelayDebounce();
+      break;
+
+    case InputDebounceType::INTERGRATING:
+      readIntegratingDebounce();
+      break;
+
+    default:
+      readNoDebounce();
+      break;
+  }
+}
+
+void DigitalInputGroup::readNoDebounce() {
+  m_values = m_inputs.read();
+}
+
+void DigitalInputGroup::readDelayDebounce() {
+  int reading = m_inputs.read();
+
+  for (unsigned i = 0; i < m_numPins; i++) {
+    // if the input changed, due to noise or actual actions (press, touch, etc.):
+    // Note if it is a noise, then the change would happen in 2x times (think of a spike, or a series of spikes)
+    // (reading & m_curr_values) gives xxxxx1/0xxxxx, then shift that to the first bit, and then check if it is 0 or 1
+    if ((~(reading & m_curr_values) >> i) & 0x1) {
+      // reset the delay start time
+      m_startTimes[i] = m_timer.read_ms();
+      ;
+
+      // set the delay started flag
+      m_delayStarted |= 1 << i;
```

Comment:
I would also include brackets around `1 << i` and similar elsewhere in code for readability.